### PR TITLE
Refactor markdown `paragraph` print

### LINKED ISSUE #9503: Inline React components in MDX cause line lengths to exceed print-width
**Prettier 2.1.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAogD3gJygQwBsAZASygGsACACxzgDMBeAHRBphgAckB6X-AHcAzgDp8AW3wAvaKMgTekACZwupLnELk4vNgD4AggHUAylQDCEVQAUNWnal5ZcBEuQr6qpYVXxVIWHIAVwhg31VtADc4HABPPyhlKnpCOHxhOD9gmAgpGFJoKkycKNIwLJgafBg-UmVfYQkICCqqVS5CCDiJBBgxKgBNMID8KG9YOABzHBqsgGYcZK58HBgE3IhCX3JcxITheC4qCAYqOLCcFK10zKouHAgK4V8IK4vg0fHwrKtbe20UCyGUSVAQygAtLkIeDilscoUoAAaCZgQjBZTkKZUADiNAghyoAEkoMJNGBcld4IdfGNksJgqRqaIQEiQBAuAVoMJkKBVo9BDZVggeSgiIJ8HEeWyAEazMAUOAwUwrMBY5AwHDBOBsuASGVwZSqZTEMZTYL4KZwABib3yBSgU2QIHwOQgrPYMAkhGMNCZcDJ+AqphFTNIZXWzrALw95BKMBssymUmQDCImTZACthJgAELyxXKyRwMhA1PpnUgbOYUxYtIARVC8HL20rKxwJWdEmUmA9DwJcGMsy4zoeAdiMQ9AEcm3BExzRS7hBCgYbDR76DPSPRE5aU0g0622ZkJKQNVrK8I63BG604C2MyAYPgZcZ6lVkAAmNma-CkQFTFYEj7iAAYAKwej8AAqL6ioej5RNqJKqLAphgDgGgwIYSSmOsaQPm2GGwG+ygfkgACMAAMlEAL40UAA)
```sh
--parser mdx
--print-width 100
```

**Input:**
```mdx
<ExternalLink href="http://aws.amazon.com/codepipeline/">AWS CodePipeline</ExternalLink> is a continuous delivery and release automation service that aids smooth deployments. You can integrate 3rd party tools into any step of your release process or you can use CodePipeline as an end-to-end solution, including Ghost Inspector tests and suites.
```

**Output:**
```mdx
<ExternalLink href="http://aws.amazon.com/codepipeline/">AWS CodePipeline</ExternalLink> is a continuous
delivery and release automation service that aids smooth deployments. You can integrate 3rd party tools
into any step of your release process or you can use CodePipeline as an end-to-end solution, including
Ghost Inspector tests and suites.

```

**Expected behavior:**

Here the output optically looks correct, however if you view it in the playground or any editor that has a line length “ruler” you can see that each line is actually longer than 100 characters.

```
❯ echo '<ExternalLink href="http://aws.amazon.com/codepipeline/">AWS CodePipeline</ExternalLink> is a continuous' | wc
       1       6     105
```

This causes a bit of a problem if you have word wrapping on in your editor.

Screen Shot 2020-10-26 at 1 58 54 PM

If I remove the inline component it works fine, so the bug seems to have something to do with inline React components.

Thanks!

### PR DESCRIPTION
**Pull request by fisker**
## Description

Fix #9503

## Checklist

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: tests/mdx/mdx/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -381,7 +382,8 @@ printWidth: 80
 </Hello>
 <Hello>
   test <World /> test
-</Hello>123
+</Hello>
+123
```
Comment (fisker) [inline tests/mdx/mdx/__snapshots__/jsfmt.spec.js.snap:386]: This change may related to #6961

**Comment by thorn0** [comment]
Ooops, I keep forgetting about this one...

**Comment by thorn0** [comment]
The doc for #9503 still doesn't look right:
```js
[
  fill([
    conditionalGroup([
      group([/*...*/]),
      group([/*...*/]),
    ]),
    " ", // <--- should be `line`
    "is",
    line,
// ...
```

File: src/language-markdown/printer-markdown.js [hunk 2]
```diff
@@ -84,15 +84,18 @@ function genericPrint(path, options, print) {
...
-    case "paragraph":
-      return printChildren(path, options, print, {
-        postprocessor: fill,
-      });
+    case "paragraph": {
+      const printed = cleanDoc(printChildren(path, options, print));
+      if (typeof printed === "string") {
+        return printed;
+      }
+      return fill(getDocParts(printed));
```
Comment (thorn0) [inline src/language-markdown/printer-markdown.js:97]: Passing the result of `cleanDoc` to `fill` is wrong. `cleanDoc` doesn't guarantee that the array has the expected structure.

**Comment by thorn0** [comment]
Yes, this is a big one. My research plan so far:

## MDX: line break inserted between elements and text in JSX blocks

```<asdf/>zxcv```

- Prettier 2.2.1 parses it all as JSX (not Markdown) and adds line breaks the way it normally does for JSX.
- That's probably not the expected output. That's why #6961 was opened.
- Also there is a `fill` bug that causes v2.2.1 to print `<Hello></Hello><Hello></Hello>123` without that break.
  ```js
  [
    fill([
      group(["<Hello", indent([]), ">"]),
      "</Hello>", // <--- Problem! Odd-indexed elements must be line breaks.
      hardlineWithoutBreakParent,
      breakParent,
      group(["<Hello", indent([]), ">"]),
      "</Hello>",
      softline,
      "123"
    ]),
    hardline,
  ]
  ```
- Things to look into:
  1. The `fill` bug
     1. Is the bug in #9503 same or different?
     2. How did it work in older versions?
     3. Was `normalizeDoc` safe before it got extracted to `doc-utils`?
     4. Manipulations with `normalizeDoc` and `cleanDoc` (in particular, PR #9847) look like they might affect more things in Markdown than just MDX. What else can be affected?
  2. What actually is the expected output?
     1. Should the text be parsed as JSX or as Markdown?

File: src/language-markdown/printer-markdown.js (hunk 2 above)
Comment (fisker) [reply src/language-markdown/printer-markdown.js:97]: Yes, you are  right, and it seems I've done something wrong in other places, I'll check.

**Pull request closed without merging**